src/realm/object-store/sync/sync_session.cpp

```diff
@@ -587,23 +587,31 @@ void SyncSession::handle_error(SyncError error)
             case ClientError::unknown_message:
             case ClientError::http_tunnel_failed:
             case ClientError::auto_client_reset_failure:
+            case ClientError::too_many_redirects:
                 // Don't do anything special for these errors.
                 // Future functionality may require special-case handling for existing
                 // errors, or newly introduced error codes.
                 break;
         }
     }
-    else {
-        // The server replies with '401: unauthorized' if the access token is invalid, expired, revoked, or the user
-        // is disabled. In this scenario we attempt an automatic token refresh and if that succeeds continue as
-        // normal. If the refresh request also fails with 401 then we need to stop retrying and pass along the error;
-        // see handle_refresh().
-        if (error_code == util::websocket::make_error_code(util::websocket::Error::bad_response_401_unauthorized)) {
+    else if (error_code.category() == util::websocket::error_category()) {
+        if (error_code == util::websocket::Error::bad_response_401_unauthorized) {
+            // If the server replies with '401: unauthorized' if the access token is invalid, expired, revoked, or the
+            // user is disabled. In this scenario we attempt an automatic token refresh and if that succeeds continue
+            // as normal. If the refresh request also fails with 401 then we need to stop retrying and pass along the
+            // error; see handle_refresh().
             if (auto u = user()) {
                 u->refresh_custom_data(handle_refresh(shared_from_this()));
                 return;
             }
         }
+        else if (error_code == util::websocket::Error::bad_response_301_moved_permanently) {
+            // If a 301 redirect is received while trying to connect to the server, the connection will attempt to
+            // connect to the server again with the updated location. Don't pass the error code.
+            return;
```

Comment:
should we also refresh the custom data here so we update the URL when the app re-opens?

Reply:
This is the `handle_error` that occurs as part of handling the redirect response from the server. The connection is set up to use the new location provided in the location field when it attempts to reconnect. I didn't want to pass this error back up to the app since a reconnect is in progress (and it's technically not an error).